hpy/debug/src/debug_ctx.c

```diff
@@ -181,6 +181,36 @@ DHPy debug_ctx_Type_FromSpec(HPyContext *dctx, HPyType_Spec *spec, HPyType_SpecP
     return DHPy_open(dctx, HPyType_FromSpec(get_info(dctx)->uctx, spec, NULL));
 }
 
+#define DEBUG_AS_STRUCT(SHAPE) \
+    void *debug_ctx_AsStruct_##SHAPE(HPyContext *dctx, DHPy dh) \
+    { \
+        HPyContext *uctx = get_info(dctx)->uctx; \
+        UHPy uh = DHPy_unwrap(dctx, dh); \
+        void *actual_data_ptr = _HPy_AsStruct_##SHAPE(uctx, uh); \
+        void *expected_data_ptr = _HPy_AsStruct_Slow(uctx, uh); \
+        if (actual_data_ptr != expected_data_ptr) { \
+            HPy_FatalError(uctx, "Invalid usage of _HPy_AsStruct_" #SHAPE \
+                ". The object's type has a different shape."); \
+        } \
```

Comment:
maybe we can be even more precise? With this logic, if two shapes happen to have the size `basicsize`, this check would pass. So ideally I would like to check that the actual shape stored in `extra->shape` matches the one which we expect.
I don't know how hard it is to access `extra->shape` from here though (if it's too hard, I'm fine with leaving this logic, but maybe with a comment which explains why it's sub-optimal and why it's too hard to do the right thing)

Reply:
Okay, I've removed `_HPyType_AsStruct_Slow` and introduced `_HPyType_GetBuiltinShape`. This is similar to CPython's function `PyObject_GetTypeDataSize` as proposed in [PEP-697](https://peps.python.org/pep-0697/).
We could also add some `_HPyType_AsStruct_FromShape(HPyContext *, HPy, HPyType_BuiltinShape)` if we need it (but that's not the case so far).
The debug mode now uses that and can do an accurate check.